instrumentation/java-classloader/src/main/java/io/opentelemetry/instrumentation/auto/javaclassloader/ResourceInjectionInstrumentation.java

```diff
@@ -0,0 +1,113 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.instrumentation.auto.javaclassloader;
+
+import static io.opentelemetry.javaagent.tooling.bytebuddy.matcher.AgentElementMatchers.extendsClass;
+import static net.bytebuddy.matcher.ElementMatchers.isMethod;
+import static net.bytebuddy.matcher.ElementMatchers.named;
+import static net.bytebuddy.matcher.ElementMatchers.takesArguments;
+
+import com.google.auto.service.AutoService;
+import io.opentelemetry.javaagent.bootstrap.HelperResources;
+import io.opentelemetry.javaagent.tooling.Instrumenter;
+import java.net.URL;
+import java.util.Collections;
+import java.util.Enumeration;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.description.method.MethodDescription;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+
+/**
+ * Instruments {@link ClassLoader} to have calls to get resources intercepted and check our map of
+ * helper resources that is filled by instrumentation when they need helpers.
+ */
+@AutoService(Instrumenter.class)
+public class ResourceInjectionInstrumentation extends Instrumenter.Default {
+
+  public ResourceInjectionInstrumentation() {
+    super("class-loader");
+  }
+
+  @Override
+  public ElementMatcher<? super TypeDescription> typeMatcher() {
+    return extendsClass(named("java.lang.ClassLoader"));
+  }
+
+  @Override
+  public Map<? extends ElementMatcher<? super MethodDescription>, String> transformers() {
+    Map<ElementMatcher<? super MethodDescription>, String> transformers = new HashMap<>();
+    transformers.put(
+        isMethod().and(named("getResource")).and(takesArguments(String.class)),
+        ResourceInjectionInstrumentation.class.getName() + "$GetResourceAdvice");
+
+    transformers.put(
+        isMethod().and(named("getResources")).and(takesArguments(String.class)),
+        ResourceInjectionInstrumentation.class.getName() + "$GetResourcesAdvice");
+    return transformers;
+  }
+
+  public static class GetResourceAdvice {
+    @Advice.OnMethodExit(suppress = Throwable.class)
+    public static void onExit(
+        @Advice.This ClassLoader classLoader,
+        @Advice.Argument(0) String name,
+        @Advice.Return(readOnly = false) URL resourceUrl) {
+      if (resourceUrl != null) {
+        // Give their classloader precedence.
+        return;
+      }
+
+      resourceUrl = HelperResources.load(classLoader, name);
+    }
+  }
```

Comment:
do we have a use case for this? this doesn't seem very auto-instrumentation-y to me (backing off if app already has a particular resource present)

Reply:
Hmm - don't think we have a use case. `getResource` won't help much in practice anyways since we're doing this mostly to interact with service loading mechanisms. It's here mostly for symmetry - on the flip side, lacking a use case, I lean towards being less intrusive into the app but what do you think?

Reply:
Without a use case, I would suggest removing instrumentation of `getResource`

Reply:
Now I am confused: how this resource injection works without this instrumentation?

Reply:
Service loading mechanisms call `getResources`, not `getResource`, since they need to find all the classes of an interface usually. So `getResource` isn't really used in practice. But @trask does this potential confusion make the case that we should stick to symmetry here?

Reply:
That should definitely be documented :)

Reply:
I'm not sure if that's the right implementation for `getResource` (singular) instrumentation, e.g. we may want to merge contents in some way, but it's hard to know without a use case, so would suggest we remove it until we need it

Reply:
Makes sense - deleted